segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1325 @@
+     * of the Target StreamSegment (but only if its length equals the given offset), after which the Source StreamSegment
+     * will cease to exist. Prior to this operation, the Source StreamSegment must be sealed.
+     *
+     * @param targetHandle  A read-write SegmentHandle that points to the Target StreamSegment. After this operation
+     *                      is complete, this is the surviving StreamSegment.
+     * @param offset        The offset in the Target StreamSegment to concat at.
+     * @param sourceSegment The Source StreamSegment. This StreamSegment will be concatenated to the Target StreamSegment.
+     *                      After this operation is complete, this StreamSegment will no longer exist.
+     * @param timeout       Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded. If the operation failed,
+     * it will contain the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> BadOffsetException: When the given offset does not match the actual length of the target segment in storage.
+     * <li> StreamSegmentNotExistsException: When the either the source Segment or the target Segment do not exist in Storage.
+     * <li> StorageNotPrimaryException: When this Storage instance is no longer primary for the target Segment (it was fenced out).
+     * <li> IllegalStateException: When the Source Segment is not Sealed.
+     * </ul>
+     * @throws IllegalArgumentException If targetHandle is read-only.
+     */
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration timeout) {
+        checkInitialized();
+        return execute( () -> {
+            long traceId = LoggerHelpers.traceEnter(log, "concat", targetHandle, offset, sourceSegment);
+            Timer timer = new Timer();
+
+            Preconditions.checkArgument(null != targetHandle, "targetHandle");
+            Preconditions.checkArgument(!targetHandle.isReadOnly(), "targetHandle");
+            Preconditions.checkArgument(null != sourceSegment, "targetHandle");
+            Preconditions.checkArgument(offset >= 0, "offset");
+            String targetSegmentName = targetHandle.getSegmentName();
+
+            // This is a critical assumption at this point which should not be broken,
+            if (null != systemJournal) {
+                Preconditions.checkState(!systemJournal.isStorageSystemSegment(targetSegmentName));
+                Preconditions.checkState(!systemJournal.isStorageSystemSegment(sourceSegment));
+            }
+
+            try (MetadataTransaction txn = metadataStore.beginTransaction()) {
+
+                // Validate preconditions.
+                SegmentMetadata targetSegmentMetadata = (SegmentMetadata) txn.get(targetSegmentName);
+                if (null == targetSegmentMetadata || !targetSegmentMetadata.isActive()) {
```

Comment:
There are many if blocks to basically throw the same exception. I wonder if this kind of logic could be compacted on a more concise way.

Reply:
done.